src/Schedgen2/mpi_colls.py

```diff
@@ -56,103 +27,185 @@ def dissemination(comm_size, datasize, tag):
     return comm
 
 
-def recdoub_allreduce(comm, comm_size, datasize, base_tag, ctd=0):
-    num_steps = int(log2(comm_size))
-    for rank in range(0, comm_size):
-        # Reduce-scatter
-        sources = [rank ^ (2**i) for i in range(num_steps)]
-        destinations = sources
-        data_sizes_receive = [datasize // (2**i) for i in range(1, num_steps + 1)]
-        data_sizes_send = data_sizes_receive
-        dependency = iterative_send_recv(
-            comm,
-            rank,
-            sources,
-            destinations,
-            data_sizes_receive,
-            data_sizes_send,
-            base_tag,
-            compute_time_dependency=ctd,
-        )
+def incast(
+    comm_size: int,
+    datasize: int,
+    tag: int = 42,
+    ptrn: str = "linear",
+    unbalanced: bool = False,
+    **kwargs,
+):
+    assert ptrn == "linear", "incast only supports the linear communication pattern"
+    return linear(
+        comm_size=comm_size,
+        datasize=datasize,
+        tag=tag,
+        algorithm="incast",
+        parallel=True,
+        randomized_data=unbalanced,
+        **kwargs,
+    )
 
-        base_tag += 1
-        # Allgather
-        sources = sources[::-1]
-        destinations = sources
-        data_sizes_receive = data_sizes_receive[::-1]
-        data_sizes_send = data_sizes_send[::-1]
-        iterative_send_recv(
-            comm,
-            rank,
-            sources,
-            destinations,
-            data_sizes_receive,
-            data_sizes_send,
-            base_tag,
-            last_dependency=dependency,
-            compute_time_dependency=ctd,
-        )
 
+def outcast(
```

Comment:
this should not be here, there is no MPI_Outcast function in the MPI standard, so there should be no outcast function in this file. please remove.

Reply:
will move to a new file